### PR TITLE
De-dupe ReconnectApp in the persisted requests queue

### PR DESCRIPTION
**Pull request by gedu**
### Details

we have 4 places that can trigger the ReconnectApp:
1. NetInfo.addEventListener
2. Middleware: [recheckConnection](https://github.com/Expensify/App/blob/main/src/libs/Middleware/RecheckConnection.ts)
3. NetworkConnection.triggerReconnectionCallbacks(‘Pusher re-subscribed to private user channel’);
4. new one above: Interval recheckNetworkConnection

so, In the PersistedRequest file, within the save function, I would add a check to see if a ReconnectApp already exists in the persistedRequest queue. If it does, replace it with the new incoming ReconnectApp. This ensures we always maintain just one instance of it, with the most up-to-date information to send.

### Fixed Issues

$ https://github.com/Expensify/App/issues/47742
PROPOSAL: https://github.com/Expensify/App/issues/47742#issuecomment-2302498516

Reference: https://github.com/Expensify/App/pull/40059

### Tests

- [x] Verify that no errors appear in the JS console

Base
1. Use the app, try to use it on different situations where you think calls to ReconnectApp can happen

With failing requests
1. Go to Settings -> Troubleshoot
2. Enable `Simulate failing network requests`
3. Go to Chat and send some messages or open different reports
4. Then go back to Settings -> Troubleshoot and disable Simulate failing network requests
5. Verify that one ReconnectApp shows up on the Network tab from DevTool

With failing requests
1. Go to Settings -> Troubleshoot
2. Enable `Force offline`
3. Go to Chat and send some messages or open different reports
4. Then go back to Settings -> Troubleshoot and disable `Force offline`
5. Verify that one ReconnectApp shows up on the Network tab from DevTool

### QA Steps

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/user-attachments/assets/6d42a8c1-9b56-43b5-9812-f95de88e6ce3

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/user-attachments/assets/cea235e9-f32e-4607-bf48-35e75556fd51

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/user-attachments/assets/685f57c7-4dfe-4892-b09c-6e4d2cec0c7f

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/user-attachments/assets/c02440ee-d413-4f0d-ac99-25e2d7b76e77

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/user-attachments/assets/ddf96eac-7565-4c68-9adf-e422a284aa31

https://github.com/user-attachments/assets/20015719-4b0f-474d-aeb8-bc850a3a0051

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/user-attachments/assets/a135e1cc-a697-44b4-ae1f-9022d6f7a9ee

</details>

### CODE REVIEW

**Comment by gedu** [comment]
@roryabraham PR created, ready for review

**Review by roryabraham — CHANGES_REQUESTED** [review]
I think we should:

- create a more generalized framework for de-duping network requests in the request queue, as is done in [this PR](https://github.com/Expensify/App/pull/40059)
- add back all the other cases that were removed (i.e: if adding `DeleteComment` to the request queue, and there's any `AddComment` or `UpdateComment` for the same reportActionID, then just remove them all from the queue).
- Add automated tests

Honestly, I was planning to do this myself because I have good historical context, but I just haven't prioritized it yet. I've finished up a lot of other stuff so might be able to proceed with https://github.com/Expensify/App/pull/40059 now.

**Comment by gedu** [comment]
@roryabraham Nice, I like that idea of making it generic, I didn't know that other requests will need some strategy to fix it. Give that many requests require different strategies, for reconnectApp the structure you made for `XXXXComment`doesn't suit well. I would like to propose a single function that does what the request needs. Like `checkAndFixConflictingRequest`, which would return
```
{
  requests: [], // requests fixed
  canSave: boolean, // if the requests sent can be saved into PersistedRequests
}
```
That way all is encapsulate into one place, into the request creation.
Maybe I can create the interface here, apply the reconnectApp strategy, add some tests. We can keep handling the `XXXXComments` in a separate PR

**Comment by roryabraham** [comment]
> for reconnectApp the structure you made for XXXXCommentdoesn't suit well

Maybe I'm not seeing something, but I think it works fine for `ReconnectApp`:

1. For `getConflictingRequests`, return any array of any other serialized `ReconnectApp` calls
1. For `shouldSkipThisRequestOnConflict`, return `false`.
1. For `handleConflictingRequest`, do nothing

**Comment by gedu** [comment]
@roryabraham I don’t want to add extra functions that aren’t needed. As an example, `bulkRemove` won’t work for the **ReconnectApp** because it deletes the instance, and the new one gets added in a different position. I want to update the requests.

Let’s say we have 10 items in the queue, and there is a **ReconnectApp** at position 4; a new one will be added. The `getConflictingRequests` will return the item from position 4. If I delete it and add the new one, it will be added at the end, resulting in a different position.
Secondly, the code:
```
const requests = PersistedRequests.getAll().filter((persistedRequest) => persistedRequest !== currentRequest);
```
won’t work for me because I need to find the position of the **ReconnectApp** to replace it with the new data, so I need the index.

I think the main logic from your PR can be inside the `checkAndFixConflictingRequest` (we can change the name), and it will work the same way. 
This function will return a flag to check if it should save or update the request:
```
if (persistAction.type === CONST.SAVE) {
    PersistedRequests.save(request);
} else if (persistAction.type === CONST.UPDATE) {
    PersistedRequests.update(persistAction.index, request);
}
```
(I'm updating a bit while thinking and finding more cases)

**Comment by gedu** [comment]
@roryabraham in my last commit It is your code, from the PR, using just one single function, and you can see how reconnectApp will work. What do you think?

**Review by roryabraham — CHANGES_REQUESTED** [review]
Overall this looks like a good implementation. It's an improvement over what I had in that it identifies the main use-cases we'll have and implements them with better guard rails. Good job 👍🏼 

NAB but I kind of prefer `replace` and `push` as the actions instead of `update` and `save`, because they're a bit more distinctive from each other and better reflect the queue structure.

Also, I'd love to add some automated tests for this before merging it. Thanks!

**Comment by gedu** [comment]
@roryabraham Great approach! It definitely makes sense. I will work on ReconnectApp with the new conflictResolver. This way, each change has its own PR, which makes it easier to track and identify any regressions. After it, we can proceed with looking into the DELETE_COMMENT. What do you think?
I will start working on tests and the suggested changes.

**Comment by roryabraham** [comment]
Sounds good 👍

**Comment by gedu** [comment]
@roryabraham I added a few tests, I'm giving a round of QA, what else I'm missing? what more tests do you think I should add?

File: src/types/onyx/Request.ts
Comment (roryabraham) [inline src/types/onyx/Request.ts:61]:
```suggestion
type ConflictRequestReplace = {
```

File: src/types/onyx/Request.ts
Comment (roryabraham) [inline src/types/onyx/Request.ts:59]:
```suggestion
 * Model of a conflict request that has to be replaced in the request queue.
```

File: src/types/onyx/Request.ts
Comment (roryabraham) [inline src/types/onyx/Request.ts:74]:
```suggestion
 * Model of a conflict request that has to be enqueued at the end of the request queue.
```

File: src/types/onyx/Request.ts
Comment (roryabraham) [inline src/types/onyx/Request.ts:76]:
```suggestion
type ConflictRequestPush = {
```

File: src/types/onyx/Request.ts
Comment (roryabraham) [inline src/types/onyx/Request.ts:84]:
```suggestion
 * Model of a conflict request that does not need to be updated or saved in the request queue.
```

File: tests/unit/SequentialQueueTest.ts
Comment (roryabraham) [inline tests/unit/SequentialQueueTest.ts:56]: Let's also validate here that the persisted request is the second one pushed.

File: tests/unit/SequentialQueueTest.ts
Comment (roryabraham) [inline tests/unit/SequentialQueueTest.ts:119]:
```suggestion
        const conflictResolver = (persistedRequests: Request[]): ConflictActionData => {
```

File: tests/unit/SequentialQueueTest.ts
Comment (roryabraham) [inline tests/unit/SequentialQueueTest.ts:38]: I think we should also add a test that queues other unrelated requests and make sure that the replaced request is in the same index as the conflicting `ReconnectApp` request.

i.e: [`OpenReport`, `AddComment`, `ReconnectApp`, `OpenReport`]

becomes [`OpenReport`, `AddComment`, `ReconnectApp`, `OpenReport`], with `ReconnectApp` still at index 2 for example

**Review by roryabraham — CHANGES_REQUESTED** [review]
It's great that we've added tests to cover the underlying logic in `SequentialQueue`, but I think it's also important that we add test coverage for `App.reconnectApp` that ensures that our implementation of `checkAndFixConflictingRequest` is working as expected in that case.

**Comment by gedu** [comment]
@roryabraham it is ready for your review

File: src/libs/Network/SequentialQueue.ts
Comment (roryabraham) [inline src/libs/Network/SequentialQueue.ts:217]: Let's add a log for the `noAction` case

**Review by roryabraham — CHANGES_REQUESTED** [review]
Still missing tests that wrap `App.reconnectApp` to ensure that the `requestConflictResolver` there is implemented correctly (and won't be broken in subsequent changes)

File: tests/unit/PersistedRequests.ts
Comment (gedu) [reply tests/unit/PersistedRequests.ts:49]: Yeah, in the beforeEach, saving a new request caught me too before. The test was already written, I'm adding more cases. I will work on the suggestions

**Comment by gedu** [comment]
@roryabraham can you take a look? I improved the tests

File: src/libs/Network/SequentialQueue.ts
Comment (roryabraham) [inline src/libs/Network/SequentialQueue.ts:28]:
```suggestion
```

File: src/libs/Network/SequentialQueue.ts
Comment (roryabraham) [inline src/libs/Network/SequentialQueue.ts:167]:
```suggestion
```

File: src/libs/ReportActionsUtils.ts
Comment (roryabraham) [inline src/libs/ReportActionsUtils.ts:59]:
```suggestion
```

File: tests/actions/SessionTest.ts
Comment (roryabraham) [inline tests/actions/SessionTest.ts:119]: Let's also add a check to make sure that the persisted request is indeed `ReconnectApp`:

```
expect(PersistedRequests.getAll()[0].command).toBe(WRITE_COMMANDS.RECONNECT_APP);
```

File: tests/actions/SessionTest.ts
Comment (roryabraham) [inline tests/actions/SessionTest.ts:112]: NAB but you can use `async`/`await` in tests to improve readability

**Review by roryabraham — CHANGES_REQUESTED** [review]
mostly LGTM, few things

**Comment by gedu** [comment]
@roryabraham fixed the comments, and some of the deprecated errors, not sure how to improve the `getParentReportAction`, looked for some examples, and found 1 PR but they merged anyway with the error

**Comment by gedu** [comment]
@roryabraham and not sure about this one
```
function getOriginalMessage<T extends ReportActionName>(reportAction: OnyxInputOrEntry<ReportAction<T>>): OriginalMessage<T> | undefined {
    if (!Array.isArray(reportAction?.message)) {
        return reportAction?.message ?? reportAction?.originalMessage;
    }
    return reportAction?.originalMessage;
}
```
because ` reportAction?.originalMessage` is deprecated, it tells you to use `getOriginalMessage` should we return an empty string?

**Comment by roryabraham** [comment]
> because  reportAction?.originalMessage is deprecated, it tells you to use getOriginalMessage should we return an empty string?

Just suppress the lint errors in `getOriginalMessage

**Comment by roryabraham** [comment]
Got it. @gedu you can fix the lint warnings on this branch by:

1. Merging main
1. Applying this diff (reasonable to suppress the lint errors in all these cases):

```diff
diff --git a/.eslintrc.js b/.eslintrc.js
index cb121953327..c9d11b97d7d 100644
--- a/.eslintrc.js
+++ b/.eslintrc.js
@@ -108,7 +108,7 @@ module.exports = {
         'plugin:you-dont-need-lodash-underscore/all',
         'plugin:prettier/recommended',
     ],
-    plugins: ['@typescript-eslint', 'jsdoc', 'you-dont-need-lodash-underscore', 'react-native-a11y', 'react', 'testing-library', 'eslint-plugin-react-compiler'],
+    plugins: ['@typescript-eslint', 'jsdoc', 'you-dont-need-lodash-underscore', 'react-native-a11y', 'react', 'testing-library', 'eslint-plugin-react-compiler', 'deprecation'],
     ignorePatterns: ['lib/**'],
     parser: '@typescript-eslint/parser',
     parserOptions: {
@@ -254,6 +254,7 @@ module.exports = {
                 },
             },
         ],
+        'deprecation/deprecation': 'off',
     },
 
     overrides: [
diff --git a/src/libs/ReportActionsUtils.ts b/src/libs/ReportActionsUtils.ts
index 3d0f3e1989a..842a51e69ed 100644
--- a/src/libs/ReportActionsUtils.ts
+++ b/src/libs/ReportActionsUtils.ts
@@ -213,8 +213,10 @@ function isActionOfType<T extends ReportActionName[]>(
 
 function getOriginalMessage<T extends ReportActionName>(reportAction: OnyxInputOrEntry<ReportAction<T>>): OriginalMessage<T> | undefined {
     if (!Array.isArray(reportAction?.message)) {
+        // eslint-disable-next-line deprecation/deprecation
         return reportAction?.message ?? reportAction?.originalMessage;
     }
+    // eslint-disable-next-line deprecation/deprecation
     return reportAction?.originalMessage;
 }
 
@@ -335,7 +337,7 @@ function isThreadParentMessage(reportAction: OnyxEntry<ReportAction>, reportID:
 /**
  * Returns the parentReportAction if the given report is a thread/task.
  *
- * @deprecated Use Onyx.connect() or withOnyx() instead
+ * @deprecated Use Onyx.connect() or useOnyx() instead
  */
 function getParentReportAction(report: OnyxInputOrEntry<Report>): OnyxEntry<ReportAction> {
     if (!report?.parentReportID || !report.parentReportActionID) {
@@ -1752,6 +1754,7 @@ export {
     getNumberOfMoneyRequests,
     getOneTransactionThreadReportID,
     getOriginalMessage,
+    // eslint-disable-next-line deprecation/deprecation
     getParentReportAction,
     getRemovedFromApprovalChainMessage,
     getReportAction,

```

**Comment by gedu** [comment]
Hey, I’m working on adding a new test for a new approach. Previously, ongoing requests weren’t saved into Onyx; they were just removed from the memory queue. Now, I’m adding a flag that allows you to decide per request whether to store it in Onyx or not. This is useful for requests we might want to process again if the user closes the app, while others, like OpenApp, don’t need to be saved because it will . I’m currently running some tests on this

File: src/libs/Network/SequentialQueue.ts
Comment (roryabraham) [inline src/libs/Network/SequentialQueue.ts:209]: NAB but you can move this inside the `if (checkAndFixConflictingRequest)` so we don't create this variable unnecessarily if it won't be used

File: src/libs/Network/SequentialQueue.ts
Comment (roryabraham) [inline src/libs/Network/SequentialQueue.ts:216]:
```suggestion
        // don't try to serialize a function
```

File: src/libs/ReportActionsUtils.ts
Comment (roryabraham) [inline src/libs/ReportActionsUtils.ts:596]: We can't safely remove this yet. I checked logs and it's still being hit. Please add it back and suppress the deprecation warning

File: src/libs/actions/PersistedRequests.ts
Comment (roryabraham) [inline src/libs/actions/PersistedRequests.ts:119]: Pretty sure this will be more efficient and avoid copying all the persisted requests.

```suggestion
    persistedRequests.unshift(ongoingRequest);
```

**Review by roryabraham — CHANGES_REQUESTED** [review]

File: src/libs/actions/PersistedRequests.ts
Comment (gedu) [reply src/libs/actions/PersistedRequests.ts:119]: Good catch, was replacing to then and forgot this one

File: .eslintrc.js
Comment (roryabraham) [inline .eslintrc.js:259]: This is a dupe now. It was added in another PR. You can remove it here:

```suggestion
```

**Review by roryabraham — DISMISSED** [review]
Otherwise I think this prettymuch LGTM 👍🏼

**Comment by gedu** [comment]
@mountiny @roryabraham how is testing going here?

**Comment by roryabraham** [comment]
> @mountiny @roryabraham how is testing going here?

At this point the PR LGTM, I'm just waiting for @shubham1206agra to review and test

**Comment by roryabraham** [comment]
conflicts btw

**Comment by mountiny** [comment]
Please use this thread to try to wrap this PR up as soon as possible https://expensify.slack.com/archives/C05LX9D6E07/p1727863372595789

**Review by shubham1206agra — APPROVED** [review]

**Review by roryabraham — APPROVED** [review]

File: src/libs/actions/PersistedRequests.ts
Comment (suneox) [inline src/libs/actions/PersistedRequests.ts:105]: Coming from #57119 checklist: After shifting persistedRequests, the request still includes the same shifted request in offline mode, like with toggle API requests. When we handle endRequestAndRemoveFromQueue it ends up removing the same shifted request. To handle this case, we need to ensure the request is unique.

**Pull request merged**
